### PR TITLE
Object selection prototype

### PR DESCRIPTION
**Pull request by paavohuhtala**
Implements object selection using a draggable selection box. The selection box is converted into a frustum (based on code from [here](https://ghoscher.com/2010/12/09/xna-picking-tutorial-part-ii/)), and intersected with every object in the scene. Selected objects are marked with rectangles.

Changelog:

* Implement both single and area object selection
* Add `IsSelectable` to `GameObject` and parse it from `MapObject`
* Add `SelectionGui`, which serves as a debug overlay for object selection. It can also serve as the model for implementing tooltips and health bars.
* Add `void Render(DrawingContext2D drawingContext)` to `Scene3D`, for rendering aforementioned UI elements.
* Remove `DebugEntityPickerSystem`
* Significantly improve `BoundingBox.Transform` to use an exact solution instead of an approximation.
* Vectorise and cleanup other functions in `BoundingBox`
* Like in the original engine, the camera can now be rotated by holding down the middle mouse button. For laptop users left ALT+Mouse1 will also work.

Known issues and limitations:

* Selection isn't as precise as on SAGE, because it's based on bounding boxes. A ray-mesh collision -based and a shader-based approach are ways of fixing this.
* The selection box can't be dragged to go "under" the control bar. Attempting this causes glitches. This is because the wnd system doesn't know about the selection system, so it consumes all mouse events when the mouse goes over it.
* Similarly, if the mouse cursor leaves the window during selection, box selection continues. This is a known limitation of the input system.
* Object selection events are not posted to`EchoConnection` and not handled by `OrderProcessor`.
* If you're using a laptop without a mosue3 button / gesture while viewing a map with the data viewer, you can no longer rotate the camera. This is because the data viewer doesn't work with keyboard events right now.

### CODE REVIEW

File: src/OpenSage.Game/Content/MapLoader.cs [hunk 1]
```diff
@@ -265,6 +265,11 @@ private static GameObject CreateGameObject(MapObject mapObject, Team[] teams, Co
                 gameObject.Owner = team;
             }
 
+            if (mapObject.Properties.TryGetValue("objectSelectable", out var selectable))
```
[the remark this answers is not in the document]
Comment (tgjones) [reply src/OpenSage.Game/Content/MapLoader.cs]: For objects that are created during gameplay, I guess they *will* need to use `GameObject.Definition.KindOf.Get(ObjectKinds.Selectable)`, in the `GameObject` constructor?

File: src/OpenSage.Game/Logic/Object/Colliders.cs [hunk 2]
```diff
@@ -1,34 +1,39 @@
...
                 // Assumes uniform scaling
-                depth *= _transform.Scale;
+                depth *= Transform.Scale;
                 return true;
             }
 
             return false;
         }
 
+        public abstract bool Intersects(in BoundingFrustum frustum);
```
[the remark this answers is not in the document]
Comment (tgjones) [reply src/OpenSage.Game/Logic/Object/Colliders.cs]: I agree.

File: src/OpenSage.Game/Logic/SelectionSystem.cs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                // Note that the box can be scaled down after this.
+                SelectionGui.SelectionBoxVisible = true;
+            }
+
+            SelectionGui.SelectionRectangle = rect;
+        }
+
+        public void OnEndDragSelection()
+        {
+            // TODO: Handle multi / single selection
```
Comment (tgjones) [inline src/OpenSage.Game/Logic/SelectionSystem.cs]: Is this TODO obsolete?

File: src/OpenSage.Game/Mathematics/BoundingBox.cs [hunk 4]
```diff
@@ -123,21 +120,67 @@ public PlaneIntersectionType Intersects(ref Plane plane)
...
+
+        /// <summary>
+        /// Converts the bounding box from world space into a screen space bounding rectangle.
+        /// </summary>
+        public Rectangle GetBoundingRectangle(CameraComponent camera)
+        {
+            var topLeft = new Vector3(float.MaxValue);
+            var bottomRight = new Vector3(float.MinValue);
+
+            unsafe
```
[the remark this answers is not in the document]
Comment (tgjones) [reply src/OpenSage.Game/Mathematics/BoundingBox.cs]: This is clever. We can come back to it once `stackalloc` can be used without `unsafe` without requiring a preview version of `System.Memory`.

File: src/OpenSage.Game/Logic/SelectionSystem.cs (hunk 3 above)
Comment (paavohuhtala) [reply src/OpenSage.Game/Logic/SelectionSystem.cs]: Yes.

**Pull request merged**